### PR TITLE
Change asset pipeline default to Propshaft in Rails 8

### PR DESCRIPTION
**Pull request by dhh**
Sprockets has served us well, but it's time to hand over the torch to [Propshaft](https://github.com/rails/propshaft) in Rails 8.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Lots of tests broken, but once we fix them, we can merge this. main is already Rails 8

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt [hunk 1]
```diff
@@ -13,10 +13,11 @@
     <link rel="icon" href="/icon.png" type="image/png">
     <link rel="icon" href="/icon.svg" type="image/svg+xml">
     <link rel="apple-touch-icon" href="/icon.png">
+    <% style_link_target = options[:asset_pipeline] == "propshaft" ? ":all" : "application" %>
```
[the remark this answers is not in the document]
Comment (dhh) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:16]: I'd say it'd usually be a feature to include css from the gem in the app, but yes, might not work for all cases. I think this is the better default, though, when we're promoting a #nobuild approach to CSS that relies on vanilla.

[the remark this answers is not in the document]
Comment (brunoprietog) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:16]: With this configuration, say, you install mission_control-jobs. Now, Mission Control's CSS conflicts with your application's CSS even on global stuff. I feel that this should not happen... What do you think about that?

[the remark this answers is not in the document]
Comment (brunoprietog) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:16]: Using `@import` for #nobuild won't be a viable option? This [PR](https://github.com/rails/propshaft/pull/175) tries to solve the existing problem in Propshaft with that. The only thing I don't like about it is that you would have to manually declare each file.

[the remark this answers is not in the document]
Comment (dhh) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:16]: I think mission-control's css should be properly scoped that this won't be a problem. But also open to looking for other solutions that do not require manually maintaining a list of CSS files to be included. Maybe mission control gets a way to opt out of :all through a configuration.

The @import issue will be fixed before Rails 8 in Propshaft. That's definitely an issue.

File: railties/lib/rails/generators/app_base.rb [hunk 2]
```diff
@@ -75,7 +75,7 @@ def self.add_shared_options_for(name)
 
         class_option :skip_asset_pipeline, type: :boolean, aliases: "-A", default: nil
 
-        class_option :asset_pipeline,      type: :string, aliases: "-a", default: "sprockets",
+        class_option :asset_pipeline,      type: :string, aliases: "-a", default: "propshaft",
```
Comment (MatheusRich) [inline railties/lib/rails/generators/app_base.rb:78]: Might be done in a follow-up PR, but are there documentation changes that we'd like to do as well?

**Comment by dhh** [comment]
There's a fair amount of work in switching all the tests that assume sprockets over. If anyone is interested in helping with this transition, I'd love some assistance, while I tune up Propshaft with the other stuff we're missing before a release.

**Comment by fedesapuppo** [comment]
I'd very much like to help with the transition. I'm not an expert, just an enthusiast. If you think it's something I could help with, I gladly offer my assistance 🚀

**Review by ItaloCobains — COMMENTED** [review]
Sounds good.

File: railties/test/generators/shared_generator_tests.rb [hunk 3]
```diff
@@ -51,7 +51,7 @@ def test_implied_options_with_conflicting_option
   end
 
   def test_skeleton_is_created
-    run_generator
+    run_generator [destination_root]
```
Comment (lazaronixon) [inline railties/test/generators/shared_generator_tests.rb:54]:
```suggestion
    run_generator
```
This is probably not required.

Comment (lazaronixon) [inline railties/test/generators/shared_generator_tests.rb:54]:
```suggestion
    run_generator
```
This is probably not required.

**Comment by dhh** [comment]
Will need to follow up with documentation changes to match the new default. But will do that in a separate PR.

**Pull request merged**